# Combining Data

Practice combining data from two different data sets. In the same folder as this Jupyter notebook, there are two csv files:
* rural_population_percent.csv
* electricity_access_percent.csv

They both come from the World Bank Indicators data. 
* https://data.worldbank.org/indicator/SP.RUR.TOTL.ZS
* https://data.worldbank.org/indicator/EG.ELC.ACCS.ZS

The rural populaton data represents the percent of a country's population that is rural over time. The electricity access data shows the percentage of people with access to electricity.

In this exercise, you will combine these two data sets together into one pandas data frame.

# Exercise 1

Combine the two data sets using the [pandas concat method](https://pandas.pydata.org/pandas-docs/stable/merging.html). In other words, find the union of the two data sets.

In [1]:
import pandas as pd

In [3]:
df_rural = pd.read_csv('rural_population_percent.csv', skiprows=3)

In [4]:
df_rural.head()

Country Name Country Code                            Indicator Name  \
0        Aruba          ABW  Rural population (% of total population)   
1  Afghanistan          AFG  Rural population (% of total population)   
2       Angola          AGO  Rural population (% of total population)   
3      Albania          ALB  Rural population (% of total population)   
4      Andorra          AND  Rural population (% of total population)   

   Indicator Code    1960    1961    1962    1963    1964    1965  ...  \
0  SP.RUR.TOTL.ZS  49.224  49.239  49.254  49.270  49.285  49.300  ...   
1  SP.RUR.TOTL.ZS  91.779  91.492  91.195  90.890  90.574  90.250  ...   
2  SP.RUR.TOTL.ZS  89.565  89.202  88.796  88.376  87.942  87.496  ...   
3  SP.RUR.TOTL.ZS  69.295  69.057  68.985  68.914  68.842  68.770  ...   
4  SP.RUR.TOTL.ZS  41.550  39.017  36.538  34.128  31.795  29.555  ...   

     2009    2010    2011    2012    2013    2014    2015    2016    2017  \
0  56.579  56.941  57.302  57.636  57.942  58.221  58.472  58.696  58.893   
1  75.687  75.311  74.926  74.532  74.129  73.718  73.297  72.868  72.430   
2  60.701  59.903  59.100  58.301  57.510  56.726  55.950  55.181  54.422   
3  48.924  47.837  46.753  45.670  44.617  43.591  42.593  41.624  40.684   
4  11.648  12.183  12.740  13.292  13.835  14.367  14.885  15.388  15.873   

   Unnamed: 62  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  

[5 rows x 63 columns]

In [6]:
df_electricity = pd.read_csv('electricity_access_percent.csv', skiprows=3)

In [8]:
df_electricity.head()

Country Name Country Code                           Indicator Name  \
0        Aruba          ABW  Access to electricity (% of population)   
1  Afghanistan          AFG  Access to electricity (% of population)   
2       Angola          AGO  Access to electricity (% of population)   
3      Albania          ALB  Access to electricity (% of population)   
4      Andorra          AND  Access to electricity (% of population)   

   Indicator Code  1960  1961  1962  1963  1964  1965  ...        2009  \
0  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   93.354546   
1  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   44.854885   
2  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...   32.382469   
3  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...  100.000000   
4  EG.ELC.ACCS.ZS   NaN   NaN   NaN   NaN   NaN   NaN  ...  100.000000   

         2010        2011        2012        2013        2014        2015  \
0   93.356292   93.942375   94.255814   94.578262   94.906723   95.238182   
1   42.700000   43.222019   69.100000   67.259552   89.500000   71.500000   
2   33.514950   34.600000   35.821964   36.990490   32.000000   42.000000   
3  100.000000  100.000000  100.000000  100.000000  100.000000  100.000000   
4  100.000000  100.000000  100.000000  100.000000  100.000000  100.000000   

         2016  2017  Unnamed: 62  
0   95.570145   NaN          NaN  
1   84.137138   NaN          NaN  
2   40.520607   NaN          NaN  
3  100.000000   NaN          NaN  
4  100.000000   NaN          NaN  

[5 rows x 63 columns]

In [11]:
df_rural.columns

Index(['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code',
       '1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', 'Unnamed: 62'],
      dtype='object')

In [12]:
df_rural = df_rural.drop(columns=['Unnamed: 62'])
df_electricity = df_electricity.drop(columns=['Unnamed: 62'])

In [17]:
pd.concat([df_rural, df_electricity]).shape, df_rural.shape, df_electricity.shape

((528, 62), (264, 62), (264, 62))

In [3]:
# TODO: import the pandas library

# TODO: read in each csv file into a separate variable
# HINT: remember from the Extract material that these csv file have some formatting issues
# HINT: The file paths are 'rural_population_percent.csv' and 'electricity_access_percent.csv'
df_rural = None
df_electricity = None

# TODO: remove the 'Unnamed:62' column from each data set

# TODO: combine the two data sets together using the concat method

# Exercise 2 (Challenge)

This exercise is more challenging.

The resulting data frame should look like this:

|Country Name|Country Code|Year|Rural_Value|Electricity_Value|
|--|--|--|--|--|--|
|Aruba|ABW|1960|49.224|49.239|
... etc.

Order the results in the dataframe by country and then by year

Here are a few pandas methods that should be helpful:
* [melt](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html)
* [drop](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)
* [merge](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.merge.html)
* [sort_values](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)

HINT: You can use country name, country code, and the year as common keys between the data sets

In [24]:
df_rural.shape

(264, 62)

In [27]:
years = ['1960', '1961', '1962', '1963', '1964', '1965', '1966', '1967', '1968',
       '1969', '1970', '1971', '1972', '1973', '1974', '1975', '1976', '1977',
       '1978', '1979', '1980', '1981', '1982', '1983', '1984', '1985', '1986',
       '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994', '1995',
       '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003', '2004',
       '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017']

df_rural_melted = pd.melt(df_rural, id_vars=['Country Name', 'Country Code'], 
        value_vars=years,var_name='Year', value_name='Rural Value').sort_values(['Country Name','Year'])

df_electricity_melted = pd.melt(df_electricity, id_vars=['Country Name', 'Country Code'], 
        value_vars=years,var_name='Year', value_name='Electricity Value').sort_values(['Country Name','Year'])

In [33]:
pd.concat?

Signature:
pd.concat(
    objs: Union[Iterable[~FrameOrSeries], Mapping[Union[Hashable, NoneType], ~FrameOrSeries]],
    axis=0,
    join='outer',
    ignore_index: bool = False,
    keys=None,
    levels=None,
    names=None,
    verify_integrity: bool = False,
    sort: bool = False,
    copy: bool = True,
) -> Union[ForwardRef('DataFrame'), ForwardRef('Series')]
Docstring:
Concatenate pandas objects along a particular axis with optional set logic
along the other axes.

Can also add a layer of hierarchical indexing on the concatenation axis,
which may be useful if the labels are the same (or overlapping) on
the passed axis number.

Parameters
----------
objs : a sequence or mapping of Series or DataFrame objects
    If a mapping is passed, the sorted keys will be used as the `keys`
    argument, unless it is passed, in which case the values will be
    selected (see below). Any None objects will be dropped silently unless
    they are all None in which case a ValueError will be raise

In [ ]:
df_combined

In [38]:
# TODO: merge the data sets together according to the instructions. First, use the 
# melt method to change the formatting of each data frame so that it looks like this:
# Country Name, Country Code, Year, Rural Value
# Country Name, Country Code, Year, Electricity Value

# TODO: drop any columns from the data frames that aren't needed

# TODO: merge the data frames together based on their common columns
# in this case, the common columns are Country Name, Country Code, and Year

# TODO: sort the results by country and then by year

df_combined = df_rural_melted.merge(df_electricity_melted, on=['Country Name', 'Country Code', 'Year'])

In [39]:
df_combined

Country Name Country Code  Year  Rural Value  Electricity Value
0      Afghanistan          AFG  1960       91.779                NaN
1      Afghanistan          AFG  1961       91.492                NaN
2      Afghanistan          AFG  1962       91.195                NaN
3      Afghanistan          AFG  1963       90.890                NaN
4      Afghanistan          AFG  1964       90.574                NaN
...            ...          ...   ...          ...                ...
15307     Zimbabwe          ZWE  2013       67.346          37.076813
15308     Zimbabwe          ZWE  2014       67.499          32.300000
15309     Zimbabwe          ZWE  2015       67.624          33.700000
15310     Zimbabwe          ZWE  2016       67.723          38.145138
15311     Zimbabwe          ZWE  2017       67.793                NaN

[15312 rows x 5 columns]

In [36]:
df_combined[df_combined['Country Name'] == 'Afghanistan']

Country Name Country Code  Year  Rural Value  \
Country Name 1      Afghanistan          AFG  1960       91.779   
             265    Afghanistan          AFG  1961       91.492   
             529    Afghanistan          AFG  1962       91.195   
             793    Afghanistan          AFG  1963       90.890   
             1057   Afghanistan          AFG  1964       90.574   
...                         ...          ...   ...          ...   
Country Code 13993  Afghanistan          AFG  2013          NaN   
             14257  Afghanistan          AFG  2014          NaN   
             14521  Afghanistan          AFG  2015          NaN   
             14785  Afghanistan          AFG  2016          NaN   
             15049  Afghanistan          AFG  2017          NaN   

                    Electricity Value  
Country Name 1                    NaN  
             265                  NaN  
             529                  NaN  
             793                  NaN  
             1057                 NaN  
...                               ...  
Country Code 13993          67.259552  
             14257          89.500000  
             14521          71.500000  
             14785          84.137138  
             15049                NaN  

[116 rows x 5 columns]